### ISSUE TITLE
Basic Auth

### ISSUE THREAD

**Issue by tanmayb123** [Alamofire/Alamofire#2542]
This isn't as much of an 'issue' as it as a 'feature request' (I'm happy to put together a pull request for this).

I'd like to make it easier for people to use basic auth with Alamofire. Instead of using a code snippet like this:
```
let user = "your username"

let password = "your password"

let credentialData = "\(user):\(password)".data(using: String.Encoding(rawValue: String.Encoding.utf8.rawValue))!

let base64Credentials = credentialData.base64EncodedString()

let headers = ["Authorization": "Basic \(base64Credentials)"]

Alamofire.request("URL", method: .get, parameters: nil, encoding: URLEncoding.default, headers: headers).responseJSON {
```
(from: https://stackoverflow.com/questions/43833743/basic-auth-header-in-alamofire --- there are many people (mainly beginners) that aren't able to figure out how to set this up)

We can do something like so:
```
Alamofire.request("URL", method: .get, basicAuth: (username: "username", password: "password"), parameters: nil, encoding: URLEncoding.default, headers: headers).responseJSON {
```

Basically, compressing all those lines of code down into one.

Is this something that would be valuable for me to implement?

**Comment by tanmayb123** [comment]
I was thinking about a function like this one (I put this together in a few minutes, so it's probably not final code):
```
/// Creates a `DataRequest` using the default `SessionManager` to retrieve the contents of the specified `url`,
/// `method`, `parameters`, `encoding` and `headers`. If `basicAuth` is provided, function will pass credentials
/// to `headers`. If `headers` already has the "Authorization" tag, the value in `headers` will take priority.
///
/// - parameter url:        The URL.
/// - parameter basicAuth:  The basic auth credentials. `nil` by default.
/// - parameter method:     The HTTP method. `.get` by default.
/// - parameter parameters: The parameters. `nil` by default.
/// - parameter encoding:   The parameter encoding. `URLEncoding.default` by default.
/// - parameter headers:    The HTTP headers. `nil` by default.
///
/// - returns: The created `DataRequest`.
@discardableResult
public func request(
    _ url: URLConvertible,
    basicAuth: (username: String, password: String)? = nil,
    method: HTTPMethod = .get,
    parameters: Parameters? = nil,
    encoding: ParameterEncoding = URLEncoding.default,
    headers: HTTPHeaders? = nil)
    -> DataRequest
{
    var auth_headers = headers
    if let authCreds = basicAuth {
        var appendAuthToHeader = false
        if let contains_auth = headers?.keys.contains("Authorization") {
            if !contains_auth {
                appendAuthToHeader = true
            }
        } else {
            auth_headers = [:]
            appendAuthToHeader = true
        }
        if appendAuthToHeader {
            let credentialData = "\(authCreds.username):\(authCreds.password)".data(using: String.Encoding(rawValue: String.Encoding.utf8.rawValue))!.base64EncodedString()
            auth_headers!["Authorization"] = "Basic \(credentialData)"
        }
    }
    return SessionManager.default.request(
        url,
        method: method,
        parameters: parameters,
        encoding: encoding,
        headers: auth_headers
    )
}
```

**Comment by tanmayb123** [comment]
Just realized Alamofire already has this, haha.
